src/sleep/sleep.c

```diff
@@ -617,6 +593,19 @@ static int run(int argc, char *argv[]) {
                                        "Sleep operation \"%s\" is disabled by configuration, refusing.",
                                        sleep_operation_to_string(arg_operation));
 
+        /* Freeze the user sessions */
+        r = getenv_bool("SYSTEMD_SLEEP_SKIP_FREEZE_USER_SESSIONS");
+        if (r < 0 && r != -ENXIO)
+                log_warning_errno(r, "Cannot parse value of $SYSTEMD_SLEEP_SKIP_FREEZE_USER_SESSIONS, ignoring.");
+        if (r <= 0) {
+                r = unit_freezer_freeze(SPECIAL_USER_SLICE, &user_slice_freezer);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to freeze user sessions, ignoring: %m");
+                else
+                        log_info("Froze user sessions");
+        } else
+                log_notice("User sessions remain unfrozen on explicit request. This is not recommended and might create deadlocks, security leaks, data loss, and other undesired behavior; especially if home directory encryption or hybrid sleep is used.");
```

Comment:
I don't think we should add such annoyance, especially when it's expected for nvidia packages to ship those drop-ins, and this message will be recorded in journal every time user does the sleep...

Please just do a generic `log_info("User session slices (not) frozen.")`.

Also, why would this "create deadlock"? I'm thinking that freezing is more likely to result in unexpected deadlocks.

Reply:
I added the big long message at request of @poettering: https://github.com/systemd/systemd/pull/30612#discussion_r1440327851

> Also, why would this "create deadlock"?

hmm I can't think of a great way to deadlock on this. @poettering ideas?

> I'm thinking that freezing is more likely to result in unexpected deadlocks

It can result in hangs, absolutely. But shouldn't result in deadlocks. It's unfrozen on resume, after all

Reply:
this message btw, would be a great fit for the ENVIRONMENT.md text above.

Reply:
message should say that the explicit request was via the env var, to give people a hint what they have to do to get back normal behaviour.

Reply:
what i meant by "deadlock" is that if we don't freeze the user's whole session then some of the user's code will end up freezing because it touches the disk and that is broken, and other stuff of the user will not directly because it doesn#t touch the disk itself, but then might hang because the user services it contacts don#t react anymore.

if we however freeze the whole session, then the problem goes away pretty much, as everything of the user is frozen then, and nothing will actively wait for anything else anymore because it itself is frozen too.

So yeah, maybe "deadlock" is a bit of a misnomer for this situation, but not sure how to say this better. the difference is really between "active hang" and "inactive hang", by which i mean that in case of an "active hang" a process runs normally except that nothing it tries responds anymore to it and thus it effectively hangs. And an "inactive hang" is that the proces itself is frozen, and thus cannot possibly be affected by hangs of the services it deals with, because it simply cannot contact them itself because it itself is frozen...

Reply:
Yeah so not deadlock

Thinking more about this: the issues come from homed, not from systemd-sleep. And I have made homed freeze the user session on lock separately. So what situations are we protecting against here?

I mean I know this is something we should be doing. For one, it standardizes the behavior of systemd-sleep across the various modes. It gives us the flexibility to do some more advanced sleep behaviors (like those from Android), where we don't actually fully suspend the system at the kernel level. It solves the weirdness of letting userspace run in the brief period of awake-ness that exists while the system transitions from suspend to hibernate. My point is I get why we need this, but I don't know how to describe it succinctly in either this message or in docs/ENVIRONMENT.md

This warning should probably go into homed, not systemd-sleep, no?

Reply:
The way I see it is this: previously there was one reason to freeze the user sessions: s2h wakeups. Now there's a 2nd one: homed. The first reason resulted in us adding a specifically conditioned freeze. The 2nd reason could theoretically result in a different specifically conditioned freeze. But maintaining these specific conditions is a brittle, and unnecessary (because freezing – in theory at least, ignoring the mess that the nvidia driver is – should not hurt anything), hence we just made it all a single un-conditioned freeze, which simplifies things.

I think systemd-sleep is still the right place for it, since it generically implements this now for all scenarios.

Reply:
maybe:

```c
log_notice("User sessions remain unfrozen on explicit request ($SYSTEMD_SLEEP_FREEZE_USER_SESSIONS is set to false). This is not not recommended, and might result in unexpected behaviour, in particular in suspend-then-hibernate operations and setups with encrypted home directories.");
```
I think something like this finds a nice balance between being vague and being specific and is still somewhat short.